### PR TITLE
Revert "Revert "Support use alias column in indices""

### LINKED ISSUE #15449: ALIAS column in secondary index usage.
**Use case**
We have some complex function, which we would like to use in secondary index, and we want to create an alias column as that function in order to have simple and memoriable name to it, which we can use in WHERE clause.

**Describe the solution you'd like**
Clickhouse server version 20.9.2.20
```
CREATE TABLE test_index
(
    `key_string` String,
    `key_uint32` ALIAS toUInt32(key_string),
    INDEX idx key_uint32 TYPE set(0) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY tuple()
PRIMARY KEY tuple()
ORDER BY key_string


Received exception from server (version 20.9.2):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'key_uint32' while processing query: 'key_uint32', required columns: 'key_uint32', source columns: 'key_string'.


CREATE TABLE test_index
(
    `key_string` String,
    `key_uint32` ALIAS toUInt32(key_string),
    INDEX idx toUInt32(key_string) TYPE set(0) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY tuple()
PRIMARY KEY tuple()
ORDER BY key_string;

INSERT INTO test_index SELECT * FROM numbers(1000000);

SELECT *
FROM test_index
WHERE key_uint32 = 1

┌─key_string─┐
│ 1          │
└────────────┘

1 rows in set. Elapsed: 0.010 sec. Processed 1.00 million rows, 14.89 MB (99.93 million rows/s., 1.49 GB/s.)

SELECT *
FROM test_index
WHERE toUInt32(key_string) = 1

┌─key_string─┐
│ 1          │
└────────────┘

1 rows in set. Elapsed: 0.008 sec. Processed 8.19 thousand rows, 121.96 KB (1.02 million rows/s., 15.20 MB/s.)
```
**Describe alternatives you've considered**
Write that kind of conditions in full form, but that makes less sense when we have the ability to use ALIAS columns.

### LINKED ISSUE #55650: Cannot use ALIAS columns in indices
This does not work:
```
CREATE TABLE test
(
    x UInt32,
    y ALIAS x + 1,
    INDEX i_y (y) TYPE minmax
) ENGINE = MergeTree ORDER BY x;
```

This does:
```
CREATE TABLE test
(
    x UInt32,
    y ALIAS x + 1,
    INDEX i_y (x + 1) TYPE minmax
) ENGINE = MergeTree ORDER BY x;
```

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #15449 
Fixes #55650

This PR re-introduces #57220 which was reverted by #57537 (4 out of 4032 runs of `02911_support_alias_column_in_indices.sql` in CI failed). The only [change](https://github.com/ClickHouse/ClickHouse/pull/57546/commits/697e967b40e1a15784aeaf5ab0f820a2352d27d3) over the original PR is that the test was stabilized.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
It is now possible to refer to ALIAS column in index (non-primary-key) definitions (issue #55650). Example: `CREATE TABLE tab(col UInt32, col_alias ALIAS col + 1, INDEX idx (col_alias) TYPE minmax) ENGINE = MergeTree ORDER BY col;`

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit dfceeac06b5e2ef6dc34e02d2d25fad79c9248f2 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57546/dfceeac06b5e2ef6dc34e02d2d25fad79c9248f2/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7105353164">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57546/dfceeac06b5e2ef6dc34e02d2d25fad79c9248f2/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: tests/queries/0_stateless/02911_support_alias_column_in_indices.sql [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+create table test1
+(
+    c UInt32,
+    a alias c + 1,
+    index i (a) type minmax
+)
+engine = MergeTree
+order by c
+settings index_granularity = 8192;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02911_support_alias_column_in_indices.sql:15]: Also index_granularity_bytes?

File: tests/queries/0_stateless/02911_support_alias_column_in_indices.sql [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+create table test2
+(
+    c UInt32,
+    a1 alias c + 1,
+    a2 alias a1 + 1,
+    index i (a2) type minmax
+)
+engine = MergeTree
+order by ch
+settings index_granularity = 8192;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02911_support_alias_column_in_indices.sql:32]: .

File: tests/queries/0_stateless/02911_support_alias_column_in_indices.sql (hunk 1 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/02911_support_alias_column_in_indices.sql:15]: missed that, thanks.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**